clustermesh: allow waiting for the CiliumClusterConfig to appear when required

giorio94 opened the pull request:
Reporting the description of the first commit for convenience:

```
857060bbcf07 ("clustermesh: Introduce CiliumClusterConfig") introduced
the CiliumClusterConfig, which the clustermesh-apiserver stores into the
kvstore during startup, and the remote cilium agents read when connecting
to new clusters. Cilium agents tolerate the possible lack of the cluster
configuration, to account for backward compatibility.

This process currently suffers from the possibility of a race condition,
because the agents might connect to the kvstore before that the cluster
configuration has been actually written. Although as of today it doesn't
raise particular concerns given that it is only used for sanity checks,
the issue will be more serious when depending on the information therein
contained to tune the behavior. One such case concerns the upcoming
kvstoremesh approach (caching the information of remote clusters in
the local kvstore), which leverages different prefixes for separation
(hence requiring the cluster config to be present for disambiguation)
and increases the race condition window (since the cluster configuration
of new clusters would be written at runtime rather than on startup).

This commit introduces a new well-known key (`cilium/.has-cilium-config`)
that, when present, conveys that the cluster configuration will be
eventually created, and remote clusters shall wait until it is present
rather than falling back to the legacy approach. This key will be
created by the clustermesh-apiserver init container, ensuring that it is
always present when etcd is exposed to the remote agents.

Alternative approaches that have been considered, but discarded because
they would solve the problem for vanilla clustermesh but not for
kvstoremesh include:
* Setting the CiliumClusterConfig in the clustermesh-apiserver init
  container rather than during the initialization phase.
* Leveraging a readiness probe to set the clustermesh-apiserver as ready
  only once the initialization phase completed, and the configuration
  has been written.
```

Another possibility might involve being strict and making the cluster configuration mandatory starting from v1.14. This looks a no-go to me though, because:
* It would prevent a cluster Cilium v1.14 to mesh with another running Cilium v1.12 (I couldn't find any official documentation about the maximum supported skew in the clustermesh scenario to say whether this is acceptable or not).
* AFAIK. we are not currently setting the cluster configuration when the clustermesh-apiserver is not used (i.e., Cilium is run with the kvstore backend). In this scenario, even a cluster running Cilium v1.13 would be incompatible (unless the cluster config is created manually before upgrading the other clusters).  

/cc @YutaroHayakawa, @marseel, @aanm, @joestringer for opinions.

<!-- Description of change -->

```release-note
clustermesh: allow waiting for the CiliumClusterConfig to appear when required
```

giorio94 committed:
clustermesh: introduce the cilium/.has-cilium-config key

857060bbcf07 ("clustermesh: Introduce CiliumClusterConfig") introduced
the CiliumClusterConfig, which the clustermesh-apiserver stores into the
kvstore during startup, and the remote cilium agents read when connecting
to new clusters. Cilium agents tolerate the possible lack of the cluster
configuration, to account for backward compatibility.

This process currently suffers from the possibility of a race condition,
because the agents might connect to the kvstore before that the cluster
configuration has been actually written. Although as of today it doesn't
raise particular concerns given that it is only used for sanity checks,
the issue will be more serious when depending on the information therein
contained to tune the behavior. One such case concerns the upcoming
kvstoremesh approach (caching the information of remote clusters in
the local kvstore), which leverages different prefixes for separation
(hence requiring the cluster config to be present for disambiguation)
and increases the race condition window (since the cluster configuration
of new clusters would be written at runtime rather than on startup).

This commit introduces a new well-known key (`cilium/.has-cilium-config`)
that, when present, conveys that the cluster configuration will be
eventually created, and remote clusters shall wait until it is present
rather than falling back to the legacy approach. This key will be
created by the clustermesh-apiserver init container, ensuring that it is
always present when etcd is exposed to the remote agents.

Alternative approaches that have been considered, but discarded because
they would solve the problem for vanilla clustermesh but not for
kvstoremesh include:
* Setting the CiliumClusterConfig in the clustermesh-apiserver init
  container rather than during the initialization phase.
* Leveraging a readiness probe to set the clustermesh-apiserver as ready
  only once the initialization phase completed, and the configuration
  has been written.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
clustermesh: configure the .has-cluster-config key in the init container

Create the `cilium/.has-cluster-config` key through the
clustermesh-apiserver init container, to ensure that it is always
present when the remote agents connect to etcd.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
clustermesh: retry to retrieve the cluster configuration when required

Leverage the newly introduced `.has-cluster-config` key to retrigger the
retrieval of the cluster configuration in case it was not found, but it
is known that it will be eventually present. This removes the
possibility of race conditions if an agent connects to the kvstore
before that the corresponding cluster configuration has been written.
Additionally, the cluster configuration can also be forced as mandatory
through a dedicated function parameter, when that is known agent-side
because an enabled function requires it to be always present.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

YutaroHayakawa commented on pkg/clustermesh/clustermesh.go:
```diff
@@ -110,6 +110,16 @@ func GetClusterConfig(clusterName string, backend kvstore.BackendOperations) (*c
 
+// IsClusterConfigRequired returns whether the remote kvstore guarantees that the
+// cilium cluster config will be eventually created.
+func IsClusterConfigRequired(backend kvstore.BackendOperations) (bool, error) {
```
Why can't we propagate `DoFunc` context to here?

giorio94 replied:
Updated. While I was there, I've also added the context propagation to the Get/SetClusterConfig helpers.

YutaroHayakawa commented on pkg/clustermesh/remote_cluster.go:
```diff
@@ -275,6 +277,53 @@ func (rc *remoteCluster) restartRemoteConnection(allocator RemoteIdentityWatcher
+	// controller backoff period to avoid recreating every time a new connection to the remote
+	// kvstore, which would introduce an unnecessary overhead. Still, we do return in case of
+	// consecutive failures, to ensure that we do not retry forever is something strange happened.
+	err = wait.PollImmediateInfiniteWithContext(ctx, clusterConfigPollInterval, func(ctx context.Context) (done bool, err error) {
```
We don't need the name for return value? At least, `done` is not used.

giorio94 replied:
Yeah, I agree that it is not strictly required, but IMHO it helps a bit clarifying the meaning of the boolean parameter.

giorio94 committed:
clustermesh: propagate context to Get/SetClusterConfig functions

Let's propagate the context from the called to these helper functions,
rather than using context.Background().

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

sayboras reviewed (approved):
The explanation in the commit message is just awesome :100:, the overall changes make sense to me :heavy_check_mark:

YutaroHayakawa reviewed (approved):
Now looks good to me. Thanks!

aanm commented on pkg/clustermesh/remote_cluster.go:
```diff
@@ -274,6 +276,53 @@ func (rc *remoteCluster) restartRemoteConnection(allocator RemoteIdentityWatcher
+	// controller backoff period to avoid recreating every time a new connection to the remote
+	// kvstore, which would introduce an unnecessary overhead. Still, we do return in case of
+	// consecutive failures, to ensure that we do not retry forever is something strange happened.
+	err = wait.PollImmediateInfiniteWithContext(ctx, clusterConfigPollInterval, func(ctx context.Context) (done bool, err error) {
```
Any reason why we don't use [our controllers](https://github.com/cilium/cilium/blob/main/pkg/controller/manager.go) for this? They have the benefit of showing up in `cilium status` if something fails. With ` wait.PollImmediateInfiniteWithContext` the failures will get hidden AFAICT.

giorio94 replied:
Essentially, the remote cluster logic is already wrapped by a controller. The idea here was to have a retry mechanism which doesn't cause that controller to restart (by simply returning the error there), because it would also cause the restart of the etcd connection, introducing an unnecessary overhead. But yeah, it makes sense to use another controller also for that rather than `wait.PollImmediateInfiniteWithContext`.

giorio94 replied:
I've updated the PR to use the controller. @aanm PTAL

aanm reviewed (approved):
Approving because my changes are not blocking but would be nice to understand the reason behind the choice of `wait.PollImmediateInfiniteWithContext`
